### PR TITLE
Direnv rewrite and security

### PR DESCRIPTION
**Pull request by samhedin**
Mainly for https://github.com/nushell/nushell/issues/1965

This PR is mainly about two things: Rewriting some ugly code and updating it to use the security features discussed in the linked issue.
This rewrite removes the use of `nu_dir_envs` in `config.toml`. Instead, directories are trusted with the use of the command `autoenv trust`. Trust can be removed with `autoenv untrust`. Both commands take an optional directory argument, if none is provided then they use the current directory.
When using the command, the content of the `.nu-env` file in the directory is hashed and saved in `nu-env.toml`, which resides next to `config.toml`. For the variables in a .nu-env to be applied, the hash of the `.nu-env` must correspond to the one saved in `nu-env.toml`.
When either of these commands is used, nushell must be restarted for the changes to apply. 
I considered adding an `autoenv reload` command, but it would kind of make the code a mess. It can probably be done if it's a highly requested feature. For now, if it's not possible to restart nushell, you can generate the `nu-env.toml` file with the hashes beforehand, and then copy it to the correct location. (maybe relevant for people who want easy config when using docker etc)

Something that remains is better errors, which I could not figure out how to do in some cases.
For example, there could be an error that mentions that a `.nu-env` file was found in the current directory, but the user needs to re-run `autoenv trust` for it to be applied. This would be sent from `directory_specific_environment.rs`, which is far away from any spans or anything that usually seems to be needed when creating a `ShellError`. Any suggestions for what I should do?
[edit] There are now errors, see comment below

There are probably quite a few more things left, but I felt like it was time to create this PR.
Note that `nushell/crates/nu-cli/tests/commands/autoenv*` are all empty to satisfy CI/CD - I do not know how to write good tests for this.
 
Two tests fail, but I'm not sure they show unexpected or unwanted behavior: 
```---- env::environment_syncer::tests::syncs_env_if_new_env_entry_in_session_is_not_in_configuration_file stdout ----
thread 'env::environment_syncer::tests::syncs_env_if_new_env_entry_in_session_is_not_in_configuration_file' panicked at 'assertion failed: `(left == right)`
  left: `[("SHELL", "/usr/bin/you_already_made_the_nu_choice"), ("anotherone", "/home/sam/github/nushell"), ("defaultenv", "/home/sam/github/nushell"), ("overwritten", "/home/sam/github/nushell"), ("homeenv", "/home/sam/github/nushell/testme/again"), ("USER", "NUNO")]`,
 right: `[("SHELL", "/usr/bin/you_already_made_the_nu_choice"), ("USER", "NUNO")]`', crates/nu-cli/src/env/environment_syncer.rs:229:13

---- env::environment_syncer::tests::nu_envs_have_higher_priority_and_does_not_get_overwritten stdout ----
thread 'env::environment_syncer::tests::nu_envs_have_higher_priority_and_does_not_get_overwritten' panicked at 'assertion failed: `(left == right)`
  left: `[("SHELL", "/usr/bin/you_already_made_the_nu_choice"), ("anotherone", "/home/sam/github/nushell"), ("defaultenv", "/home/sam/github/nushell"), ("overwritten", "/home/sam/github/nushell"), ("homeenv", "/home/sam/github/nushell/testme/again")]`,
 right: `[("SHELL", "/usr/bin/you_already_made_the_nu_choice")]`', crates/nu-cli/src/env/environment_syncer.rs:299:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I am not sure what I did to make git think I want to merge 100+ commits, since many of them are already merged. Apologies for the mess.

### CODE REVIEW

**Comment by samhedin** [comment]
Update: There are now errors for when a .nu-env file is found but not trusted, when it's malformed, and more!
They get a bit... intense sometimes as they'll show every time you do anything. Maybe that's fine, better than failing silently.
error

**Comment by jonathandturner** [comment]
@samhedin - I think at some point we'll want to add a way of logging warnings, too, that we can automatically quiet for the user if they've already seen it. Out of scope for this PR, but something we can definitely think about.

File: crates/nu-cli/tests/commands/autoenv.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (jonathandturner) [inline crates/nu-cli/tests/commands/autoenv.rs:1]: Noticed a few empty files. Were you thinking of adding some tests for trust/untrust as part of this?

**Comment by jonathandturner** [comment]
I think the PR looks good. Since it's a pretty big change, once you're ready to land I'll do some testing by hand also, and then we can get it in.

I commented that it looked like there were some files meant for testing but were empty.

Just let me know when you're ready and I'll take another look.

**Comment by samhedin** [comment]
As I wrote in my initial post, CI/CD required that those files exist. I'm not sure how to test this code, so I sneakily added some empty files to make it quiet.

Testing this is possible. Need to add some stage to CI/CD to create some `.nu-envs` and trust them, and then check. A bit of a project though.

The last thing, which I am working on now, is to change hashing function. After that it's all ready to be tried out :)

**Comment by samhedin** [comment]
Okay, it should be ready to try out now. It uses sha256 from https://docs.rs/sha2/0.9.1/sha2/ for hashing.

**Comment by jonathandturner** [comment]
Looks like it's working. I was able to do `autoenv trust` and `autoenv untrust` and they worked well.

I think one thing that we should explore is if we can do trust/untrust and not have to restart the shell. I was a bit confused when I did `autoenv untrust` but still could see the environment, then remembered I'd have to restart the shell to see them.

Is it possible to make it a bit more automatic so that we don't have to restart?

**Comment by samhedin** [comment]
> Is it possible to make it a bit more automatic so that we don't have to restart?
I think so. The reason a restart is required is because I only
wanted to read `.config/nu/config.toml` once on startup, and
not once per directory change. However, I didn't consider
the possibility of re-reading it
if you run one of the autoenv commands.
I will think a bit about it, I agree
that it would be much nicer if it just applied immediately.
Jonathan Turner writes:

> Looks like it's working. I was able to do `autoenv trust` and `autoenv untrust` and they worked well.
>
> I think one thing that we should explore is if we can do trust/untrust and not have to restart the shell. I was a bit confused when I did `autoenv untrust` but still could see the environment, then remembered I'd have to restart the shell to see them.
>
> Is it possible to make it a bit more automatic so that we don't have to restart?

**Comment by samhedin** [comment]
Actually, nevermind. I think I know how to do it nicely now. More updates on the way!

**Comment by samhedin** [comment]
Closing for now, new and nicer one coming when it's ready. :)

**Pull request closed without merging**